File: linear_search/lsearch.rb
def find_pos(n, arr)
    arr.each_with_index do |el, i|
        return i if el == n
    end
    -1
end

def find_pos_rec(n, arr)
    return -1 if arr.empty?
    return arr.length if arr.pop() == n
    find_pos_rec(n, arr)
end

raise "First test is wrong" unless find_pos(1, [1, 7, 2, 20, 0]) == 0
raise "Second test is wrong" unless find_pos(20, [1, 7, 2, 20, 0]) == 3
raise "Third test is wrong" unless find_pos(0, [1, 7, 2, 20, 0]) == 4
raise "Fourth test is wrong" unless find_pos(-1, [1, 7, 2, 20, 0]) == -1

raise "First test is wrong" unless find_pos_rec(1, [1, 7, 2, 20, 0]) == 0
raise "Second test is wrong" unless find_pos_rec(20, [1, 7, 2, 20, 0]) == 3
raise "Third test is wrong" unless find_pos_rec(0, [1, 7, 2, 20, 0]) == 4
raise "Fourth test is wrong" unless find_pos_rec(-1, [1, 7, 2, 20, 0]) == -1

File: linear_search/lsearch.php
<?php

/**
 * complexity O(N)
 */
function findPos($n, $arr) {
    for($i = 0, $len = count($arr); $i < $len; $i++) {
        if($n === $arr[$i]) break;
    }

    if($i === $len) return -1;
    return $i;
}

function findPosRec($n, $arr) {
    if(empty($arr)) return -1;
    if($cur = array_pop($arr) === $n) return count($arr);
    return findPosRec($n, $arr);
}


$arr = [1, 4, 12, 3, 2];

assert(2 == findPos(12, $arr));
assert(0 == findPos(1, $arr));
assert(4 == findPos(2, $arr));
assert(-1 == findPos(7, $arr));

assert(2 == findPosRec(12, $arr));
assert(0 == findPosRec(1, $arr));
assert(4 == findPosRec(2, $arr));
assert(-1 == findPosRec(7, $arr));
